source/opt/inst_bindless_check_pass.cpp

```diff
@@ -87,70 +138,144 @@ void InstBindlessCheckPass::GenBindlessCheckCode(
-    load_inst = get_def_use_mgr()->GetDef(load_id);
+  return 0;
+}
+
+bool InstBindlessCheckPass::AnalyzeDescriptorReference(Instruction* ref_inst) {
+  image_id_ = GetDescriptorValueId(ref_inst);
+  if (image_id_ == 0) return false;
+  image_inst_ = get_def_use_mgr()->GetDef(image_id_);
+  if (image_inst_->opcode() == SpvOp::SpvOpSampledImage) {
+    load_id_ =
+        image_inst_->GetSingleWordInOperand(kSpvSampledImageImageIdInIdx);
+    load_inst_ = get_def_use_mgr()->GetDef(load_id_);
+  } else if (image_inst_->opcode() == SpvOp::SpvOpImage) {
+    load_id_ =
+        image_inst_->GetSingleWordInOperand(kSpvImageSampledImageIdInIdx);
+    load_inst_ = get_def_use_mgr()->GetDef(load_id_);
   } else {
-    load_id = image_id;
-    load_inst = image_inst;
-    image_id = 0;
+    load_id_ = image_id_;
+    load_inst_ = image_inst_;
+    image_id_ = 0;
   }
-  if (load_inst->opcode() != SpvOp::SpvOpLoad) {
-    // TODO(greg-lunarg): Handle additional possibilities
-    return;
+  if (load_inst_->opcode() != SpvOp::SpvOpLoad) {
+    // TODO(greg-lunarg): Handle additional possibilities?
+    return false;
   }
-  uint32_t ptr_id = load_inst->GetSingleWordInOperand(kSpvLoadPtrIdInIdx);
-  Instruction* ptr_inst = get_def_use_mgr()->GetDef(ptr_id);
-  if (ptr_inst->opcode() != SpvOp::SpvOpAccessChain) return;
-  if (ptr_inst->NumInOperands() != 2) {
-    assert(false && "unexpected bindless index number");
-    return;
+  ptr_id_ = load_inst_->GetSingleWordInOperand(kSpvLoadPtrIdInIdx);
+  ptr_inst_ = get_def_use_mgr()->GetDef(ptr_id_);
+  if (ptr_inst_->opcode() == SpvOp::SpvOpVariable) {
+    index_id_ = 0;
+    var_id_ = ptr_id_;
+    var_inst_ = ptr_inst_;
+  } else if (ptr_inst_->opcode() == SpvOp::SpvOpAccessChain) {
+    if (ptr_inst_->NumInOperands() != 2) {
+      assert(false && "unexpected bindless index number");
+      return false;
+    }
+    index_id_ = ptr_inst_->GetSingleWordInOperand(kSpvAccessChainIndex0IdInIdx);
+    var_id_ = ptr_inst_->GetSingleWordInOperand(kSpvAccessChainBaseIdInIdx);
+    var_inst_ = get_def_use_mgr()->GetDef(var_id_);
+    if (var_inst_->opcode() != SpvOpVariable) {
+      assert(false && "unexpected bindless base");
+      return false;
+    }
+  } else {
+    // TODO(greg-lunarg): Handle additional possibilities?
+    return false;
   }
-  uint32_t index_id =
-      ptr_inst->GetSingleWordInOperand(kSpvAccessChainIndex0IdInIdx);
-  ptr_id = ptr_inst->GetSingleWordInOperand(kSpvAccessChainBaseIdInIdx);
-  ptr_inst = get_def_use_mgr()->GetDef(ptr_id);
-  if (ptr_inst->opcode() != SpvOpVariable) {
-    assert(false && "unexpected bindless base");
-    return;
+  ref_inst_ = ref_inst;
+  return true;
+}
+
+void InstBindlessCheckPass::GenCheckCode(
+    uint32_t check_id, uint32_t error_id, uint32_t length_id,
+    uint32_t stage_idx, InstructionBuilder* builder,
+    std::unique_ptr<BasicBlock>* new_blk_ptr,
+    std::vector<std::unique_ptr<BasicBlock>>* new_blocks) {
```

Comment:
I have a preference for keeping the interface of a function as simple as possible.  In this case, you pass in `new_blk_ptr` only to add it to `new_blocks`.  You do nothing else with the input value.  You seem to want to do a number of things in sequence.  Add all of the instructions to a block, and then add it to `new_blocks`.  You want `new_blk_ptr` to always contain the current block our are working on.

Why couldn't you add the new blocks to `new_blocks` as soon as they are created?  Use `new_blocks->back()` in places you want to access `new_blk_ptr`.  This removes a parameter, and there is one less variable to think about.

Also, you can create the InstructionBuilder because you have to block available, right?  I know constructing and destructing can take extra time, but I don't think it is too expensive, and will simplify the interface.

Reply:
Done.